API authentication

lopesnuno opened the pull request.

lopesnuno committed:
Added authentication to endpoints

lopesnuno committed:
Added authentication to endpoints;
Added 'generate-token' directory;

lopesnuno committed:
Merge branch 'master' into api-auth

# Conflicts:
#	src/modules/message/MessageApi.ts

lopesnuno committed:
Delete keys from root folder;
Changed path to keys;

kimile commented on generate-token/index.ts:
```diff
@@ -0,0 +1,23 @@
+import { sign, SignOptions } from 'jsonwebtoken';
```
Se é para colocar no repositório:
1.  renomeia a pasta para `tools`
2.  renomeia o script para `generate-token.ts`
3. Altera o script para usar argumentos da linha de comandos ( exemplo: https://gist.github.com/kimile/aa09aada33ee06f6df3e5a04635c6ca7)
4. Cria uma entrada no package.json, secção scripts, com um comando para executar o script (https://github.com/lopesnuno/chat/blob/master/package.json#L9);  Comando deve ser `npx ts-node tools/generate-token.ts`

kimile commented on package.json:
```diff
@@ -30,7 +30,7 @@
     "dotenv": "16.0.0",
     "express": "4.17.3",
-    "jsonwebtoken": "8.5.1",
+    "jsonwebtoken": "^8.5.1",
```
Não sei porque passaste de não ter o `^` para passar a ter (deve ter sido automático) - o `^` significa que o npm pode fazer update da versão do `jsonwebtoken` para uma versão compatível (se é ou não compatível é definido por quem mantem o `jsonwebtoken`).

Ora, como nós estamos a usar o docker o sempre que fizermos um deployment de uma nova versão da API vamos correr o npm install que vai instalar tudo de raíz, ao usar o `^` correríamos o risco de de um deployment para o outro termos versão diferentes de várias dependências, e que nós não testamos de forma alguma, o que irá inevitavelmente (eventualmente) introduzir erros sem que nos apercebamos.

Remove o `^`, volta a correr o npm install para gerar o package-lock.json com a versão bloqueada àquela específica e, aproveitando a onda, cria um .npmrc na raíz do projeto com o conteúdo seguinte:
```
registry=https://registry.npmjs.org/
save-exact=true
package-lock=true
```

kimile commented on generate-token/index.js:
```diff
@@ -0,0 +1,23 @@
+exports.__esModule = true;
+exports.generateToken = void 0;
+var jsonwebtoken_1 = require("jsonwebtoken");
+var fs = require("fs");
```
Este ficheiro não deve ser incluido no projeto! Remove (vê o comentário do ficheiro .ts - usando o ts-node, não irá haver um .js generado)

kimile commented on src/modules/message/MessageApi.ts:
```diff
@@ -1,24 +1,26 @@
+    const { content, senderId, recipientId, replyTo, roomId } = req.body;
+    const id = Random.id();
+
+    await service.create(id, content, senderId, recipientId, replyTo, roomId);
```
O senderId aqui deveria ser o req.user.id, não deveria vir de quem faz o request. Assim garantes que alguém autenticado não se consegue fazer passar por outro e enviar uma mensagem em nome dele (exemplo: tu descobres o meu id, fazes um request na consola e defines o senderId como o meu id... como está autenticado, a API autoriza, e crias uma mensagem em meu nome).

kimile commented on src/modules/roomMembers/RoomMembersApi.ts:
```diff
@@ -20,5 +21,5 @@ async function deleteUser(req: Request, res: Response, next: NextFunction): Prom
 
 export default (app: Router): void => {
-    app.delete('/roomMembers/', deleteUser);
+    app.delete('/roomMembers/', Auth.authorize([]),deleteUser);
```
Não sei se estás a fazer isso no service, mas tens de garantir que o user que está a fazer esta chamada é o owner do room (fizeste isso no createRoom da roomAPI)

kimile reviewed (commented):
Genericamente, além dos comentários que coloquei, tens de rever quem está a fazer o quê, agora que sabemos quem é o utilizador que está a fazer o request.

Assim a olho:
- Actualizar/apagar um room só pode ser feito por quem o criou
- Adicionar/remover um utilizador a um room só pode ser feito por quem criou o room
- O delete de uma mensagem só pode ser feito porque a criou
- Na API de utilizadores por enquanto será melhor ignorar, porque ainda é preciso definir como vamos gerir os utilizadores na plataforma

kimile commented:
Esqueci-me de uma coisa - no teu tsconfig.json, deves ter de alterar o código ali pela linha 29 para corresponder a este snippet, para e:
```
  },
  "include": [
    "./src/**/*",
    "./tools/**/*"
  ],
```

lopesnuno committed:
- Added authentication to endpoints;
- Fixed syntax errors;
- Deleted information from payload;
- Change directory `generate-token` => `tools` and `index.ts` => `generate-token.ts`;
- Changed RoomMembers endpoint;

lopesnuno committed:
- Fixed generate-token args;
- Room return;

lopesnuno committed:
Merge branch 'master' into api-auth

# Conflicts:
#	src/modules/roomMembers/RoomMembersApi.ts

kimile commented on src/modules/roomMembers/RoomMembersRepository.ts:
```diff
@@ -27,7 +27,7 @@ export default class RoomMembersRepository implements Repository<RoomMember> {
                 DELETE
                 FROM room_members
-                WHERE id = ${id}
+                WHERE user_id = ${id}
```
Se fizeres só isto, vais remover o utilizador de todas as salas, quando o que tu deverias estar a fazer era remover apenas de uma só sala...

kimile commented on src/utils/jwt.utils.ts:
```diff
@@ -0,0 +1,35 @@
+import * as fs from 'fs';
+import * as path from 'path';
+
+export function payload() {
+    const payload = {
+        name: '',
+        userId: ''
+    };
+
+    return sign(payload);
+}
```
Para que serve isto??

lopesnuno committed:
- Changed types from delete;
- Changed delete query;
- Deleted function payload() => 'jwt.utils.ts' and sign import ;

lopesnuno committed:
Merge branch 'master' into api-auth

# Conflicts:
#	package-lock.json
#	package.json
#	src/modules/message/MessageApi.ts
#	src/modules/roomMembers/RoomMembersApi.ts
#	src/modules/roomMembers/RoomMembersRepository.ts
#	src/modules/rooms/RoomApi.ts
#	src/modules/users/UserApi.ts
#	yarn.lock

lopesnuno committed:
- Syntax fixes;
- Added .prettierrc;

lopesnuno committed:
- get function;
